spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindBadEndOfStreamCheck.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * FindBugs - Find bugs in Java programs
+ * Copyright (C) 2003,2004 University of Maryland
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+package edu.umd.cs.findbugs.detect;
+
+import org.apache.bcel.Const;
+import org.apache.bcel.classfile.JavaClass;
+
+import edu.umd.cs.findbugs.BugAccumulator;
+import edu.umd.cs.findbugs.BugInstance;
+import edu.umd.cs.findbugs.BugReporter;
+import edu.umd.cs.findbugs.OpcodeStack;
+import edu.umd.cs.findbugs.ba.XMethod;
+import edu.umd.cs.findbugs.bcel.OpcodeStackDetector;
+
+public class FindBadEndOfStreamCheck extends OpcodeStackDetector {
+
+    private final BugAccumulator bugAccumulator;
+    private OpcodeStack.Item itemUnderCast;
+    private OpcodeStack.Item castedItem;
+    private XMethod source;
+
+    public FindBadEndOfStreamCheck(BugReporter bugReporter) {
+        this.bugAccumulator = new BugAccumulator(bugReporter);
+    }
+
+    @Override
+    public void visitAfter(JavaClass obj) {
+        bugAccumulator.reportAccumulatedBugs();
+    }
+
+    @Override
+    public void sawOpcode(int seen) {
+        itemUnderCast = null;
+
+        if (seen == Const.I2B || seen == Const.I2C) {
+            OpcodeStack.Item item = stack.getStackItem(0);
+            XMethod method = item.getReturnValueOf();
+            if (method == null || !"read".equals(method.getName())
+                    || !("Ljava/io/FileInputStream;".equals(method.getClassDescriptor().getSignature())
```

Comment:
Could extract `method.getClassDescriptor().getSignature()` to a local variable. Possibly even extract all this logic to a helper like `isFileRead(Method method)`.